### PR TITLE
Notifications counter

### PR DESCRIPTION
**Pull request by sagaidak**

### CODE REVIEW

File: src/subscribe-substrate/lib/postgres.ts [hunk 1]
```diff
@@ -595,11 +601,60 @@ export const deleteCommentActivityWithActivityStream = async (userId: string,com
...
     const res = await pool.query(query, params)
     console.log(res.rows)
   } catch (err) {
     console.log(err.stack);
   }
 }
+
+export const updateUnreadNotifications = async (account: string, activityId: number) => {
+  const defaultUnreadCount = 1;
+  const query = `
```
Comment (siman) [inline src/subscribe-substrate/lib/postgres.ts]: Need to refactor this query with @siman

File: src/subscribe-substrate/lib/postgres.ts [hunk 2]
```diff
@@ -595,11 +601,60 @@ export const deleteCommentActivityWithActivityStream = async (userId: string,com
...
   try {
     const res = await pool.query(query, params)
     console.log(res.rows)
   } catch (err) {
     console.log(err.stack);
   }
 }
+
+export const updateUnreadNotifications = async (account: string, activityId: number) => {
+  const defaultUnreadCount = 1;
```
Comment (siman) [inline src/subscribe-substrate/lib/postgres.ts]: Why `1`, not `0`?

File: src/subscribe-substrate/lib/postgres.ts [hunk 3]
```diff
@@ -595,11 +601,60 @@ export const deleteCommentActivityWithActivityStream = async (userId: string,com
...
+            ) 
+            AND aggregated = true
+        )
+        last_read_activity_id = $2
+  `
+
+  const currentUnreadCount = await getUnreadNotifications(account) || 0
+  const params = [ account, activityId, defaultUnreadCount ]
+  try {
+    const res = await pool.query(query, params)
```
Comment (siman) [inline src/subscribe-substrate/lib/postgres.ts]: Why we call `await getUnreadNotifications()` before `await pool.query(query, params)`?
In this way we get old value for unread notifications, don't we?

File: src/subscribe-substrate/lib/postgres.ts [hunk 4]
```diff
@@ -595,11 +601,60 @@ export const deleteCommentActivityWithActivityStream = async (userId: string,com
...
+    console.log('Done from updateUnreadNotifications:', res.rows)
+    eventEmitter.emit('notificationUpdate', account, currentUnreadCount);
+  } catch (err) {
+    console.log('Error from updateUnreadNotifications:', err.stack);
+  }
+}
+
+export const getUnreadNotifications = async (account: string) => {
```
Comment (siman) [inline src/subscribe-substrate/lib/postgres.ts]: This func looks ok.

File: .env [hunk 5]
```diff
@@ -1,5 +1,5 @@
 # Substrate node config
-SUBSTRATE_URL=ws://127.0.0.1:9944
+SUBSTRATE_URL=ws://testnet.subsocial.network:9944
```
Comment (siman) [inline .env]: We should delete .env file from offchain.

File: src/subscribe-substrate/lib/postgres.ts (hunk 2 above)
Comment (sagaidak) [reply src/subscribe-substrate/lib/postgres.ts]: We use this number only if we adding new line to notification counter. This happens when we should update counter but there is no row with needed account.

File: src/subscribe-substrate/lib/postgres.ts (hunk 3 above)
Comment (sagaidak) [reply src/subscribe-substrate/lib/postgres.ts]: true

File: .env (hunk 5 above)
Comment (sagaidak) [reply .env]: Should I use same solution as on UI?

File: src/express-api/server.ts [hunk 6]
```diff
@@ -93,7 +93,58 @@ app.get('/v1/offchain/notifications/:id', async (req: express.Request, res: expr
   }
 });
 
-const port = 3001;
+app.get('/v1/offchain/notifications/:id/readAll', async (req: express.Request, res: express.Response) => {
```
Comment (siman) [inline src/express-api/server.ts]:
```suggestion
app.post('/v1/offchain/notifications/:id/readAll', async (req: express.Request, res: express.Response) => {

```

File: src/subscribe-substrate/lib/postgres.ts (hunk 1 above)
Comment (siman) [reply src/subscribe-substrate/lib/postgres.ts]: // Delete `defaultUnreadCount` coz we can set `1` directly in SQL query.

// Change SQL params:
// const params = [ account, activityId, defaultUnreadCount ]
// -> 
// const params = [ account ]

Try to update a query to this:

```sql
INSERT INTO df.notifications_counter 
  (account, last_read_activity_id, unread_count)
VALUES ($1, 0, 1)
ON CONFLICT (account) DO UPDATE
SET unread_count = (
  SELECT DISTINCT COUNT(*)
  FROM df.notifications
  WHERE account = $1 AND activity_id >= (
    SELECT last_read_activity_id
    FROM df.notifications_counter
    WHERE account = $1
  )
)
```

File: src/express-api/server.ts [hunk 7]
```diff
@@ -93,7 +93,58 @@ app.get('/v1/offchain/notifications/:id', async (req: express.Request, res: expr
   }
 });
 
-const port = 3001;
+app.get('/v1/offchain/notifications/:id/readAll', async (req: express.Request, res: express.Response) => {
+  const currentUnreadCount = 0
+  const { account } = req.params;
+  const query = `
+    UPDATE df.notifications_counter
+    SET unread_count = $2
```
Comment (siman) [inline src/express-api/server.ts]:
```suggestion
    SET unread_count = 0 AND last_read_notification = ___

```

File: src/express-api/server.ts [hunk 8]
```diff
@@ -93,7 +93,58 @@ app.get('/v1/offchain/notifications/:id', async (req: express.Request, res: expr
   }
 });
 
-const port = 3001;
+app.get('/v1/offchain/notifications/:id/readAll', async (req: express.Request, res: express.Response) => {
+  const currentUnreadCount = 0
```
Comment (siman) [inline src/express-api/server.ts]:
```suggestion
```

File: src/express-api/server.ts [hunk 9]
```diff
@@ -93,7 +93,58 @@ app.get('/v1/offchain/notifications/:id', async (req: express.Request, res: expr
...
 
-const port = 3001;
+app.get('/v1/offchain/notifications/:id/readAll', async (req: express.Request, res: express.Response) => {
+  const currentUnreadCount = 0
+  const { account } = req.params;
+  const query = `
+    UPDATE df.notifications_counter
+    SET unread_count = $2
+    WHERE account = $1`;
+  const params = [ account, currentUnreadCount ];
```
Comment (siman) [inline src/express-api/server.ts]:
```suggestion
  const params = [ account ];

```

File: src/sql/initDb.sql [hunk 10]
```diff
@@ -25,18 +25,25 @@ END$$;
 CREATE TABLE IF NOT EXISTS df.news_feed
 (
     account varchar(48) NOT NULL,
-    activity_id bigint NOT NULL
+    activity_id bigint NOT NULL UNIQUE
```
Comment (siman) [inline src/sql/initDb.sql]: We don't need `UNIQUE` here, multiple `account`s can have notifications about the same `activity_id`.
```suggestion
    activity_id bigint NOT NULL
```

File: src/sql/initDb.sql [hunk 11]
```diff
@@ -25,18 +25,25 @@ END$$;
...
+);
+
+CREATE TABLE IF NOT EXISTS df.notifications_counter
+(
+    account varchar(48) NOT NULL UNIQUE,
+    last_read_activity_id bigint,
```
Comment (siman) [inline src/sql/initDb.sql]:
```suggestion
    last_read_activity_id bigint DEFAULT NULL,
```

File: src/sql/initDb.sql [hunk 12]
```diff
@@ -25,18 +25,25 @@ END$$;
...
     account varchar(48) NOT NULL,
-    activity_id bigint NOT NULL
+    activity_id bigint NOT NULL UNIQUE
 );
 
 CREATE TABLE IF NOT EXISTS df.notifications
 (
     account varchar(48) NOT NULL,
-    activity_id bigint NOT NULL
+    activity_id bigint NOT NULL UNIQUE
```
Comment (siman) [inline src/sql/initDb.sql]: We don't need `UNIQUE` here, multiple `account`s can have notifications about the same `activity_id`.
```suggestion
    activity_id bigint NOT NULL
```

File: src/express-api/server.ts [hunk 13]
```diff
@@ -83,17 +83,73 @@ app.get('/v1/offchain/notifications/:id', async (req: express.Request, res: expr
...
+    console.log(data.rows);
+    res.json(data.rows);
+  } catch (err) {
+    console.log(err.stack);
+  }
+});
+
+app.post('/v1/offchain/notifications/:id/readAll', async (req: express.Request, res: express.Response) => {
+  const account = req.params.id;
+  console.log(`Deleting unread_count for ${account}`)
```
Comment (siman) [inline src/express-api/server.ts]:
```suggestion
  console.log(`Mark all notifications as read for account ${account}`)

```

File: src/express-api/server.ts [hunk 14]
```diff
@@ -83,17 +83,73 @@ app.get('/v1/offchain/notifications/:id', async (req: express.Request, res: expr
...
+  ws.on('message', async (account: string) => {
+    console.log('Received from client: %s', account);
+    const currentUnreadCount = await getUnreadNotifications(account)
+
+    clients[account] = ws;
+    clients[account].send(`${currentUnreadCount}`)
+  });
+
+  eventEmitter.on(EVENT_UPDATE_NOTIFICATIONS_COUNTER, (account: string, currentUnreadCount: number) => {
+    if (!clients[account]) return
```
Comment (siman) [inline src/express-api/server.ts]:
```suggestion
    if (!clients[account]) return

```

File: src/express-api/server.ts [hunk 15]
```diff
@@ -83,17 +83,73 @@ app.get('/v1/offchain/notifications/:id', async (req: express.Request, res: expr
...
+    if (clients[account].readyState !== WebSocket.OPEN) {
+      delete clients[account]
+      return
+    }
```
Comment (siman) [inline src/express-api/server.ts]:
```suggestion
    }

```

**Pull request merged**